### ISSUE TITLE
Hidden Select element: Select element must have an accessible name

### ISSUE THREAD

**Issue by tmc-ee** [dequelabs/axe-core#4383]
### Product

axe-core

### Product Version

4.9.0

### Latest Version

- [X] I have tested the issue with the latest version of the product

#### Expectation

A select element that is hidden should not trigger a violation based on the [select-name](https://dequeuniversity.com/rules/axe/4.8/select-name?application=axeAPI) rule, i.e. `Select element must have an accessible name`.

#### Actual

A violation is triggered for hidden select elements.

#### How to Reproduce

1. Open https://alphagov.github.io/accessible-autocomplete/examples/form.html
2. Run axe
3. The [select-name](https://dequeuniversity.com/rules/axe/4.8/select-name?application=axeAPI) violation is triggered twice (for the two hidden select elements, i.e. elements with classes `last-location` and `passport-location`)

#### Additional context

Complete tool output (`inapplicable`, `passes` and `incomplete` nodes redacted):

```
{
  "testEngine": {
    "name": "axe-core",
    "version": "4.9.0"
  },
  "testRunner": {
    "name": "axe"
  },
  "testEnvironment": {
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36",
    "windowWidth": 471,
    "windowHeight": 1035,
    "orientationAngle": 0,
    "orientationType": "landscape-primary"
  },
  "timestamp": "2024-03-26T14:17:55.247Z",
  "url": "https://alphagov.github.io/accessible-autocomplete/examples/form.html",
  "toolOptions": {
    "runOnly": {
      "type": "tag",
      "values": [
        "best-practice",
        "wcag2a",
        "wcag2aa",
        "wcag2aaa",
        "wcag21a",
        "wcag21aa",
        "wcag22aa"
      ]
    },
    "reporter": "v1"
  },
  "violations": [
    {
      "id": "select-name",
      "impact": "critical",
      "tags": [
        "cat.forms",
        "wcag2a",
        "wcag412",
        "section508",
        "section508.22.n",
        "TTv5",
        "TT5.c",
        "EN-301-549",
        "EN-9.4.1.2",
        "ACT"
      ],
      "description": "Ensures select element has an accessible name",
      "help": "Select element must have an accessible name",
      "helpUrl": "https://dequeuniversity.com/rules/axe/4.9/select-name?application=axeAPI",
      "nodes": [
        {
          "any": [
            {
              "id": "implicit-label",
              "data": null,
              "relatedNodes": [],
              "impact": "critical",
              "message": "Form element does not have an implicit (wrapped) <label>"
            },
            {
              "id": "explicit-label",
              "data": null,
              "relatedNodes": [],
              "impact": "critical",
              "message": "Form element does not have an explicit <label>"
            },
            {
              "id": "aria-label",
              "data": null,
              "relatedNodes": [],
              "impact": "critical",
              "message": "aria-label attribute does not exist or is empty"
            },
            {
              "id": "aria-labelledby",
              "data": null,
              "relatedNodes": [],
              "impact": "critical",
              "message": "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty"
            },
            {
              "id": "non-empty-title",
              "data": {
                "messageKey": "noAttr"
              },
              "relatedNodes": [],
              "impact": "critical",
              "message": "Element has no title attribute"
            },
            {
              "id": "presentational-role",
              "data": null,
              "relatedNodes": [],
              "impact": "critical",
              "message": "Element's default semantics were not overridden with role=\"none\" or role=\"presentation\""
            }
          ],
          "all": [],
          "none": [],
          "impact": "critical",
          "html": "<select name=\"last-location\" id=\"last-location-select\" style=\"display: none;\">",
          "target": [
            "#last-location-select"
          ],
          "failureSummary": "Fix any of the following:\n  Form element does not have an implicit (wrapped) <label>\n  Form element does not have an explicit <label>\n  aria-label attribute does not exist or is empty\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\n  Element has no title attribute\n  Element's default semantics were not overridden with role=\"none\" or role=\"presentation\""
        },
        {
          "any": [
            {
              "id": "implicit-label",
              "data": null,
              "relatedNodes": [],
              "impact": "critical",
              "message": "Form element does not have an implicit (wrapped) <label>"
            },
            {
              "id": "explicit-label",
              "data": null,
              "relatedNodes": [],
              "impact": "critical",
              "message": "Form element does not have an explicit <label>"
            },
            {
              "id": "aria-label",
              "data": null,
              "relatedNodes": [],
              "impact": "critical",
              "message": "aria-label attribute does not exist or is empty"
            },
            {
              "id": "aria-labelledby",
              "data": null,
              "relatedNodes": [],
              "impact": "critical",
              "message": "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty"
            },
            {
              "id": "non-empty-title",
              "data": {
                "messageKey": "noAttr"
              },
              "relatedNodes": [],
              "impact": "critical",
              "message": "Element has no title attribute"
            },
            {
              "id": "presentational-role",
              "data": null,
              "relatedNodes": [],
              "impact": "critical",
              "message": "Element's default semantics were not overridden with role=\"none\" or role=\"presentation\""
            }
          ],
          "all": [],
          "none": [],
          "impact": "critical",
          "html": "<select name=\"passport-location\" id=\"passport-location-select\" style=\"display: none;\">",
          "target": [
            "#passport-location-select"
          ],
          "failureSummary": "Fix any of the following:\n  Form element does not have an implicit (wrapped) <label>\n  Form element does not have an explicit <label>\n  aria-label attribute does not exist or is empty\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\n  Element has no title attribute\n  Element's default semantics were not overridden with role=\"none\" or role=\"presentation\""
        }
      ]
    }
  ]
}
```

**Comment by straker** [comment]
Thanks for the issue. I'm unable to reproduce it though. When I run axe-core@4.9.0 on the page it returns `select-name` as inapplicable.

<img width="622" alt="output of axe-core running only select-name rule where the result is inapplicable" src="https://github.com/dequelabs/axe-core/assets/2433219/3afaf966-f805-433f-81e0-b7bf3012b8ad">

**Comment by tmc-ee** [comment]
@straker TYVM for the quick response! Thanks to your screenshot we noticed a slight difference in the way we were running the tool. We were initially cloning the document and passing that as context to the run function, rather than just running the tool directly on the page. Running the tool directly on the page solved the issue (and removed a couple of checks that were marked `incompatible` too :raised_hands:).

I've closed this issue, below is some before and after in case others run into similar issues in future :slightly_smiling_face: 

### Before

```
async function getAxeResults() {
  const axeContext = document.cloneNode(true);

  const axeOptions: axe.RunOptions = {
    runOnly: {
      type: 'tag',
      values: [
        'best-practice',
        'wcag2a',
        'wcag2aa',
        'wcag2aaa',
        'wcag21a',
        'wcag21aa',
        'wcag22aa',
      ],
    },
  };

  const axeResults = await new Promise((resolve) => {
    axe.run(axeContext, axeOptions, (err, results) => {
      if (err) throw err;
      resolve(results);
    });
  });

  return axeResults;
}
```

### After

```
async function getAxeResults() {
  const axeOptions: axe.RunOptions = {
    runOnly: {
      type: 'tag',
      values: [
        'best-practice',
        'wcag2a',
        'wcag2aa',
        'wcag2aaa',
        'wcag21a',
        'wcag21aa',
        'wcag22aa',
      ],
    },
  };

  const axeResults = await new Promise((resolve) => {
    axe.run(axeOptions, (err, results) => {
      if (err) throw err;
      resolve(results);
    });
  });

  return axeResults;
}
```
